--- a/drivers/net/wireless/rt2x00/rt2800pci.c
+++ b/drivers/net/wireless/rt2x00/rt2800pci.c
@@ -1096,6 +1096,8 @@ static irqreturn_t rt2800pci_interrupt(i
 static int rt2800pci_read_eeprom(struct rt2x00_dev *rt2x00dev)
 {
 	int retval;
+	u8 *mac;
+	const u8 *pdata_mac;
 
 	if (rt2x00_is_soc(rt2x00dev) ||
 	    test_bit(REQUIRE_EEPROM_FILE, &rt2x00dev->cap_flags))
@@ -1105,6 +1107,11 @@ static int rt2800pci_read_eeprom(struct
 	else
 		retval = rt2800pci_read_eeprom_pci(rt2x00dev);
 
+	pdata_mac = rt2x00lib_get_mac_address(rt2x00dev);
+	mac = rt2x00_eeprom_addr(rt2x00dev, EEPROM_MAC_ADDR_0);
+	if (pdata_mac)
+		memcpy(mac, pdata_mac, ETH_ALEN);
+
 	return retval;
 }
 
