--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
+++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
@@ -103,6 +103,19 @@ static struct board_info __initdata boar
 #endif
 
 /*
+ * known 3380 boards
+ */
+#ifdef CONFIG_BCM63XX_CPU_3380
+static struct board_info __initdata board_bcm93380 = {
+	.name				= "BCM93380",
+	.expected_cpu_id		= 0x3380,
+
+	.has_uart0			= 1,
+	.has_uart1			= 1,
+};
+#endif
+
+/*
  * known 6318 boards
  */
 #ifdef CONFIG_BCM63XX_CPU_6318
@@ -4779,6 +4792,9 @@ static const struct board_info __initcon
 #ifdef CONFIG_BCM63XX_CPU_3368
 	&board_cvg834g,
 #endif
+#ifdef CONFIG_BCM63XX_CPU_3380
+	&board_bcm93380,
+#endif
 #ifdef CONFIG_BCM63XX_CPU_6318
 	&board_96318ref,
 	&board_96318ref_p300,
